README.md
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/NationalLibraryOfNorway/digital_tekstanalyse/HEAD)
# DHLAB

<!-- start dhlab-intro -->

 [`dhlab`](https://pypi.org/project/dhlab/) is a python library for doing qualitative and quantitative analyses of the digital texts from [*Nettbiblioteket*](https://www.nb.no/search) (eng: "the online library") at the [National Library of Norway](https://www.nb.no/) (NLN). *Nettbiblioteket* is the NLN's digital collection of media publications.

<!-- end dhlab-intro -->

Check out our [documentation](https://dhlab.readthedocs.io/en/latest/) for more info.

<!-- start installation -->
## Installation with pip

Install the latest version of [`dhlab`](https://pypi.org/project/dhlab/) in your (Unix) terminal with pip:

```shell
pip install -U dhlab
```
<!-- end installation -->

## Install dhlab from github repo

Open you terminal in the file location you will work with [DHLAB](https://github.com/NationalLibraryOfNorway/DHLAB).

``` shell
git clone https://github.com/NationalLibraryOfNorway/DHLAB.git
cd DHLAB
pip install -U -e .
```

## Contact
<!-- start contact-info -->
The code here is developed and maintained by [The Digital Humanities lab group](https://www.nb.no/dh-lab/).

If you have any questions, or run into any problems with the code, please log them in our [issue
tracker](https://github.com/NationalLibraryOfNorway/DHLAB/issues).
<!-- end contact-info -->


CHANGELOG.md
# Changelog

## v2.28.0 (2023-11-07)

### Feat

- **conc**: add tests for Corpus and Concordance

### Fix

- **corpus**: bug in corpus.add

### Refactor

- **corpus**: formatting

## v2.27.1 (2023-11-02)

### Fix

- **corpus**: small bug

## v2.27.0 (2023-11-02)

### Feat

- **corpus**: subsetting methods

## v2.26.5 (2023-10-19)

### Fix

- Update corpus_conc_coll.py

## v2.26.4 (2023-09-26)

### Fix

- add test placeholder

## v2.26.3 (2023-09-26)

### Fix

- add test for metadata_from_urn

## v2.26.2 (2023-09-15)

### Fix

- metadata natbib
- readme

## v2.26.1 (2023-04-12)

### Fix

- **counts**: bug on from_df method

## v2.26.0 (2023-03-16)

### Feat

- **future**: ngram + fix (#161)

## v2.25.0 (2023-03-03)

## v2.24.0 (2023-03-03)

### Feat

- konkordanser basert på ord er lagt til

## v2.23.3 (2023-03-01)

### Fix

- **counts**: (#158)

## v2.23.2 (2023-02-24)

### Fix

- **Counts**: (#157)

## v2.23.1 (2023-02-24)

### Fix

- **future**: func names (#156)

## v2.23.0 (2023-02-24)

### Feat

- **future**: (#154)

## v2.22.2 (2023-02-21)

### Fix

- **natbib, Counts**: add functools.wraps (#150)

## v2.22.1 (2023-02-16)

## v2.22.0 (2023-02-16)

## v2.21.2 (2023-02-16)

### Fix

- **freq**: add freq Counts method to corpus

## v2.21.1 (2023-02-15)

### Fix

- **Counts**: fix counts

## v2.21.0 (2023-02-13)

## v2.20.0 (2023-02-09)

### Feat

- **metadata**: update metadata service tools

## v2.19.0 (2023-01-25)

### Feat

- added urncount - removed wordcloud message

## v2.18.1 (2023-01-23)

## v2.18.0 (2023-01-18)

### Feat

- **text**: (#133)

## v2.17.1 (2023-01-16)

### Fix

- parameter passing wildcard

## v2.17.0 (2023-01-16)

### Feat

- added wildcard search for words

## v2.16.0 (2023-01-16)

### Feat

- **heatmap**: add heatmap wrapper to viz package

## v2.15.1 (2023-01-16)

### Fix

- EmptyCorpus (#129)

## v2.15.0 (2023-01-16)

### Feat

- **corpus**: (#128)

### Fix

- **bump-version**: disable triggering new workflow [skip-ci] (#127)

## v2.14.0 (2023-01-13)

### Feat

- **visualize, ngram, wordbank**: (#126)

## v2.13.3 (2023-01-09)

### Fix

- handle API errors

## v2.13.2 (2023-01-06)

### Fix

- **dhlab_api**: construct df from response.json (#124)

## v2.13.1 (2022-12-08)

### Fix

- Pos0 (#123)

## v2.13.0 (2022-12-07)

### Feat

- **api**: added start page and to page for NER and POS (#122)

## v2.12.0 (2022-12-07)

### Feat

- **ngram**: make ngram subclass of dhlabobj (#121)


## v2.11.2 (2022-11-24)

### Fix

- **ngram**: lang param is not supported for avis
- **requirements**: add scipy to reqirements

## v2.11.1 (2022-11-14)

### Feat

- **totals**: Add totals() to top level
- **DhlabObj**: add DhlabObj
- **Corpus**: allow empty Corpus and construct df

### Fix

- **conc_coll**: fix __repr__ functions

### Refactor

- **corpus**: change add()

## v2.10.0 (2022-10-26)

### Feat

- added image search in bokhylla books to api

## v2.9.5 (2022-10-26)

### Fix

- nbtokenizer now handles numbers combined with words, e.g. '1800-tallet'

## v2.9.4 (2022-10-04)

### Fix

- relative and absolute

## v2.9.3 (2022-10-04)

### Fix

- ngram smooth param

## v2.9.2 (2022-10-03)

### Fix

- lang added to ngram

## v2.9.1 (2022-09-12)

### Fix

- revert ngram_conv()

## v2.9.0 (2022-09-05)

### Feat

- added reference for words

## v2.8.2 (2022-09-01)

### Fix

- **corpus**: : extend_from_identifiers (#88)

    need to be able to work with dataframe with urns

## v2.8.1 (2022-08-30)

### Fix

- delete whitespace

## v2.8.0 (2022-08-30)

### Feat

- added spacy pos parse (pos, lemma, dependency)

### Refactor

- **corpus**: ignore index in corpus.add (#84)

## v2.7.0 (2022-08-24)

- automatic bump with features from 2.6.x versions

## v2.6.3 (2022-08-23)

### Feat

- add Empty Corpus to dhlab.__init__

### Fix

- dewey search in corpus builder

## v2.6.2 (2022-08-17)

### Feat

- add corpus default limit
- add docstring to concordance

### Fix

- revert gitignore
- revert Collocation

## v2.6.1 (2022-08-15)

### Fix

- error in pandas corpus (#75)

## v2.6.0 (2022-08-08)

### Feat

- code for word evaluations

## v2.5.0 (2022-08-05)

### Feat

-  geolocation

## v2.4.0 (2022-07-12)

### Feat

- ner with spaCy

## v2.3.6 (2022-07-12)

### Fix

- nb_ngram to point to new endpoint


## v2.3.0 (2022-06-02)

### Feat

- added access to Norsk Ordbank, wordbank

## v2.2.0 (2022-05-13)

### Feat

- ngram, geodata


## v2.1.0 (2022-05-13)

### Feat

- geodata

## v2.0.25 (2022-04-27)

### Fix

- **setup.cfg**: make package dhlab importable

## v2.0.24 (2022-04-19)

### Refactor

- expose dhlab v1 modules

## v2.0.22 (2022-03-22)

### Fix

- import all legacy modules in `__init__.py`

### Refactor
- move dhlab_v1 code into its own subpackage
- **docs/package_summary.rst**: add reference table for legacy code

## v2.0.21 (2022-03-21)

### Refactor

- **constants**: add global variables for URLs in constants.py
- Reformat code with pep8 tools
- turn relative imports into absolute imports
- simplify and reduce expressions
- rename classes with CamelCase

### Docs

- **README**: add "Example use"
- add docstrings in subpackages
- add docs/CHANGELOG.md
- **docs**: add `*.rst` documentation files
- add autosummary of whole dhlab package
- **logo**: update logo image
- add jupyter integration and toggle feature
- add copybutton to code blocks
- add docstrings and make functions private

## v2.0.20geo (2022-03-02)

### Feat

- **dhlab.api.dhlab_api**: add function `get_places`
- **text.geo_data**: add class `GeoData`

### Fix

- **text.dispersion**: pass **kwargs to `plot()`

## v.2.0.18dispersion (2022-02-21)

### Feat

- **text.dispersion**: add class Dispersion
- **api.dhlab_api**: add get_dispersion

### Fix

- **requirements**: remove wordcloud

## v2.0.17params (2022-02-08)

### Refactor

- **text.corpus**: add parameter fulltext
- **api.dhlab_api.document_corpus**: add parameter fulltext
- **text.conc_coll.Concordance**: add parameters window and limit

dhlab_v2.py
import requests
import pandas as pd
import json
import networkx as nx

BASE_URL = "https://api.nb.no/dhlab"
BASE_URL1 = "https://api.nb.no/dhlab"

pd.options.display.max_rows = 100


import re

# convert cell to a link
def make_link(row):
        r = "<a target='_blank' href = 'https://urn.nb.no/{x}'>{x}</a>".format(x = str(row))
        return r

# find hits a cell
find_hits = lambda x: ' '.join(re.findall("<b>(.+?)</b", x))

# fetch metadata

def get_metadata(urns = None):
    """ Fetch metadata from a list of urns """
    params = locals()
    r = requests.post(f"{BASE_URL}/get_metadata", json = params)
    return pd.DataFrame(r.json())


# class for displaying concordances
class Concordance:
    """Wrapper for concordance function with added functionality"""
    def __init__(self, corpus, query):
        self.concordance = concordance(urns = list(corpus.urn), words = query)
        self.concordance['link'] = self.concordance.urn.apply(make_link)
        self.concordance = self.concordance[['link', 'urn', 'conc']]
        self.concordance.columns = ['link', 'urn', 'concordance']
        self.corpus = corpus
        self.size = len(self.concordance)
    
    def show(self, n = 10, style = True):
        if style:
            result =  self.concordance.sample(min(n, self.size))[['link', 'concordance']].style
        else:
            result =  self.concordance.sample(min(n, self.size))
        return result
    
class Cooccurence():
    """Collocations """
    def __init__(self, corpus = None, words = None, before = 10, after = 10, reference = None):
        if isinstance(words, str):
            words = [words]
        coll = pd.concat([urn_collocation(urns = list(corpus.urn), word = w, before = before, after = after) for w in words])[['counts']]
        self.coll = coll.groupby(coll.index).sum()
        self.reference = reference
        self.before = before
        self.after = after
        
        if reference is not None:
            self.coll['relevance'] = (self.coll.counts/self.coll.counts.sum())/(self.reference.freq/self.reference.freq.sum())
    
    def show(self, sortby = 'counts', n = 20):
        return self.coll.sort_values(by = sortby, ascending = False)
    
    def keywordlist(self, top = 200, counts = 5, relevance = 10):
        mask = self.coll[self.coll.counts > counts]
        mask = mask[mask.relevance > relevance]
        return list(mask.sort_values(by = 'counts', ascending = False).head(200).index)
    
class Ngram():
    def __init__(self, words = None, from_year = None, to_year = None, doctype = None, lang = 'nob'):
        from datetime import datetime
        
        self.date = datetime.now()
        if to_year is None:
            to_year = self.date.year
        if from_year is None:
            from_year = 1950
            
        self.from_year = from_year
        self.to_year = to_year
        self.words = words
        self.lang = lang
        if not doctype is None:
            doctype = 'bok'
            if 'bok' in doctype:
                doctype = 'bok'
            elif 'avis' in doctype:
                doctype = 'avis'
        else:
            doctype = 'bok'
        ngrm = nb_ngram(terms = ', '.join(words), corpus = doctype, years = (from_year, to_year))
        ngrm.index = ngrm.index.astype(str)
        self.ngram = ngrm
        return None

    def plot(self, **kwargs):
        self.ngram.plot(**kwargs)
    
    def compare(self, another_ngram):
        from datetime import datetime
        start_year = max(datetime(self.from_year,1,1), datetime(another_ngram.from_year,1,1)).year
        end_year = min(datetime(self.to_year,1,1), datetime(another_ngram.to_year,1,1)).year
        compare =  (self.ngram.loc[str(start_year):str(end_year)].transpose()/another_ngram.ngram[str(start_year):str(end_year)].transpose().sum()).transpose()
        return compare

class Ngram_book(Ngram):
    """"""

    def __init__(self, words = None, title = None, publisher = None, city = None, lang = 'nob', from_year = None, to_year = None, ddk = None, subject = None):
        from datetime import datetime

        
        self.date = datetime.now()
        if to_year is None:
            to_year = self.date.year
        if from_year is None:
            from_year = 1950
        self.from_year = from_year
        self.to_year = to_year
        self.words = words
        self.title = title
        self.publisher = publisher
        self.city = city
        self.lang = lang
        self.ddk = ddk
        self.subject = subject
        self.ngram = ngram_book(word = words, title = title, publisher = publisher, lang = lang,city = city, period = (from_year, to_year), ddk = ddk, topic = subject)
        #self.cohort =  (self.ngram.transpose()/self.ngram.transpose().sum()).transpose()
        return None
    

    
class Ngram_news(Ngram):
        def __init__(self, words = None, title = None, city = None, from_year = None, to_year = None):
            from datetime import datetime


            self.date = datetime.now()
            if to_year is None:
                to_year = self.date.year
            if from_year is None:
                from_year = 1950
            self.from_year = from_year
            self.to_year = to_year
            self.words = words
            self.title = title
            self.ngram = ngram_news(word = words, title = title, period = (from_year, to_year))
            #self.cohort =  (self.ngram.transpose()/self.ngram.transpose().sum()).transpose()
            return None

def get_reference(corpus = 'digavis', from_year = 1950, to_year = 1955, lang = 'nob', limit = 100000):
    params = locals()
    r = requests.get(BASE_URL + "/reference_corpus", params = params)
    if r.status_code == 200:
        result = r.json()
    else:
        result = []
    return pd.DataFrame(result, columns = ['word', 'freq']).set_index('word')

def find_urns(docids = None, mode = 'json'):
    """ Return a list of URNs from a list of docids as a dictionary {docid: URN} or as a pandas dataframe"""
    
    params = locals()
    r = requests.post(BASE_URL1 + "/find_urn", json = params)
    if r.status_code == 200:
        res = pd.DataFrame.from_dict(r.json(), orient = 'index', columns = ['urn'])
    else:
        res = pd.DataFrame()
    return res

def ngram_book(word = ['.'], title = None, period = None, publisher = None, lang=None, city = None, ddk = None, topic = None):
    """Get a time series for a word as string, title is name of book period is (year, year), lang is three letter iso code.
    Use % as wildcard where appropriate - no wildcards in word and lang"""
    params = locals()
    if isinstance(word, str):
        # assume a comma separated string
        word = [w.strip() for w in word.split(',')]
    params['word'] = tuple(word)
    params = {x:params[x] for x in params if not params[x] is None}
    r = requests.post(BASE_URL1 + "/ngram_book", json = params)
    #print(r.status_code)
    df = pd.DataFrame.from_dict(r.json(), orient = 'index')
    df.index = df.index.map(lambda x: tuple(x.split()))
    columns = df.index.levels[0]
    df = pd.concat([df.loc[x] for x in columns], axis = 1)
    df.columns = columns 
    #df.index = df.index.map(pd.Timestamp)
    return df

def ngram_periodicals(word = ['.'], title = None, period = None, publisher = None, lang=None, city = None, ddk = None, topic = None):
    """Get a time series for a word as string, title is name of periodical period is (year, year), lang is three letter iso code.
    Use % as wildcard where appropriate - no wildcards in word and lang"""
    params = locals()
    if isinstance(word, str):
        # assume a comma separated string
        word = [w.strip() for w in word.split(',')]
    params['word'] = tuple(word)
    params = {x:params[x] for x in params if not params[x] is None}
    r = requests.post(BASE_URL1 + "/ngram_periodicals", json = params)
    #print(r.status_code)
    df = pd.DataFrame.from_dict(r.json(), orient = 'index')
    df.index = df.index.map(lambda x: tuple(x.split()))
    columns = df.index.levels[0]
    df = pd.concat([df.loc[x] for x in columns], axis = 1)
    df.columns = columns 
    #df.index = df.index.map(pd.Timestamp)
    return df


def ngram_news(word = ['.'], title = None, period = None):
    """ get a time series period is a tuple of (year, year), (yearmonthday, yearmonthday) 
    word is string and title is the title of newspaper, use % as wildcard"""
    params = locals()
    if isinstance(word, str):
        # assume a comma separated string
        word = [w.strip() for w in word.split(',')]
    params['word'] = tuple(word)
    params = {x:params[x] for x in params if not params[x] is None}
    r = requests.post(BASE_URL1 + "/ngram_newspapers", json = params)
    #print(r.status_code)
    df = pd.DataFrame.from_dict(r.json(), orient = 'index')
    df.index = df.index.map(lambda x: tuple(x.split()))
    columns = df.index.levels[0]
    df = pd.concat([df.loc[x] for x in columns], axis = 1)
    df.columns = columns
    #df.index = df.index.map(pd.Timestamp)
    return df

def get_document_frequencies(urns = None, cutoff = 0, words = None):
    params = locals()
    r = requests.post(BASE_URL1 + "/frequencies", json = params)
    result = r.json()
    structure = {u[0][0] : dict([tuple(x[1:]) for x in u]) for u in result if u != []}
    df = pd.DataFrame(structure)
    return df.sort_values(by = df.columns[0], ascending = False)

def get_document_corpus(**kwargs):
    return document_corpus(**kwargs)

def document_corpus(doctype = None, author = None, freetext = None, from_year = None, to_year = None, from_timestamp = None, to_timestamp = None, title = None, ddk = None, subject = None, lang = None, limit = None):
    """ Fetch a corpus based on metadata - doctypes are digibok, digavis, digitidsskrift"""
    
    parms = locals()
    params = {x:parms[x] for x in parms if not parms[x] is None }
    if "ddk" in params:
        params["ddk"]  = "^" + params['ddk'].replace('.', '"."')
        
    r=requests.post(BASE_URL + "/build_corpus", json=params)
    
    return pd.DataFrame(r.json())
    
def urn_collocation(urns = None, word = 'arbeid', before = 5, after = 0, samplesize = 200000):
    """ Create a collocation from a list of URNs - returns distance (sum of distances and bayesian distance) and frequency"""
    params = {
        'urn': urns,
        'word': word,
        'before': before,
        'after': after,
        'samplesize': samplesize
    }
    r = requests.post(BASE_URL1 + "/urncolldist_urn", json = params)
    return pd.read_json(r.text)

def totals(n = 50000):
    """ Get total frequencies of words in database"""
    
    r = requests.get(BASE_URL + "/totals/{n}".format(n = n))
    return pd.DataFrame.from_dict(dict(r.json()),orient = 'index', columns = ['freq'])
    
def concordance(urns = None, words = None, window = 25, limit = 100):
    """ Get a list of concordances from database, words is an fts5 string search expression"""
    if words is None:
        return {}
    else:
        params = {
            'urns': urns,
            'query': words,
            'window': window,
            'limit': limit
        }
        r = requests.post(BASE_URL + "/conc", json = params)
    return pd.DataFrame(r.json())

def concordance_counts(urns = None, words = None, window = 25, limit = 100):
    """ Get a list of concordances from database, words is an fts5 string search expression"""
    if words is None:
        return {}
    else:
        params = {
            'urns': urns,
            'query': words,
            'window': window,
            'limit': limit
        }
        r = requests.post(BASE_URL + "/conccount", json = params)
    return pd.DataFrame(r.json())

def konkordans(urns = None, query = None, window = 25, limit = 100):
    if query is None:
        return {}
    else:
        params = {
            'urns': urns,
            'query': query,
            'window': window,
            'limit': limit
        }
        r = requests.post(BASE_URL + "/conc", json = params)
    return pd.DataFrame(r.json())

def collocation(corpusquery = 'norge', word = 'arbeid', before = 5, after = 0):
    params = {
        'metadata_query': corpusquery,
        'word': word,
        'before': before,
        'after': after
    }
    r = requests.post(BASE_URL1 + "/urncolldist", json = params)
    return pd.read_json(r.text)



### -------------------------- NB ngram ------------------ ###

def nb_ngram(terms, corpus='bok', smooth=3, years=(1810, 2010), mode='relative'):
    df = ngram_conv(get_ngram(terms, corpus=corpus), smooth=smooth, years=years, mode=mode)
    df.index = df.index.astype(int)
    return df

def get_ngram(terms, corpus='avis'):
    req = requests.get(
        "https://beta.nb.no/dhlab/ngram_1/ngram/query?", 
        params = { 
            'terms':terms,
            'corpus':corpus
        }
    )
    if req.status_code == 200:
        res = req.text
    else:
        res = "[]"
    return json.loads(res)

def ngram_conv(ngrams, smooth=1, years=(1810,2013), mode='relative'):
    ngc = {}
    # check if relative frequency or absolute frequency is in question
    if mode.startswith('rel') or mode=='y':
        arg = 'y'
    else:
        arg = 'f'
    for x in ngrams:
        if x != []:
            ngc[x['key']] = {z['x']:z[arg] for z in x['values'] if int(z['x']) <= years[1] and int(z['x']) >= years[0]}
    return pd.DataFrame(ngc).rolling(window=smooth, win_type='triang').mean()


def make_word_graph(words, corpus = 'all', cutoff = 16, leaves = 0):
    """Get galaxy from ngram-database. 
    corpus is bok, avis or both
    words is a commaseparated string
    English and German provided by Google N-gram. 
    Set leaves=1 to get the leaves. Parameter cutoff only works for lang='nob'. 
    Specify English by setting lang='eng' and German by lang='ger'"""
    
    params = dict()
    params['terms'] = words
    params['corpus'] = corpus
    params['limit'] = cutoff
    params['leaves'] = leaves
    result = requests.get("https://beta.nb.no/dhlab/galaxies/query", params=params)
    G = nx.DiGraph()
    edgelist = []
    if result.status_code == 200:
        graph = json.loads(result.text)
        #print(graph)
        nodes = graph['nodes']
        edges = graph['links']
        for edge in edges:
            edgelist += [(nodes[edge['source']]['name'], nodes[edge['target']]['name'], abs(edge['value']))]
    #print(edgelist)
    G.add_weighted_edges_from(edgelist)

    return G